### PR TITLE
Parallel cooccurrence counting

### PR DESCRIPTION
**Pull request by skystrife**
This PR parallelizes the cooccurrence counting application for the GloVe embedding method using the new `corpus::parallel_consume` functionality.

It also fixes the somewhat hilarious typo of `coocur` and replaces it with `cooccur` like I actually intended.

### CODE REVIEW

File: include/meta/embeddings/cooccurrence_counter.h [hunk 1]
```diff
@@ -0,0 +1,225 @@
+/**
+ * @file coocurrence_counter.h
```
Comment (smassung) [inline include/meta/embeddings/cooccurrence_counter.h]: typo

File: include/meta/embeddings/cooccurrence_counter.h [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+    }
+
+    uint64_t bytes_read() const
+    {
+        return sizeof(count_type) * idx_;
+    }
+
+    bool operator==(const memory_cooccur_iterator& other) const
+    {
+        return items_.empty() && other.items_.empty();
```
Comment (smassung) [inline include/meta/embeddings/cooccurrence_counter.h]: So it's never possible that `items_` is nonempty and equal to `other.items_`?

File: include/meta/util/multiway_merge.h [hunk 3]
```diff
@@ -84,14 +85,21 @@ namespace util
  *     A unary function that is called once per every unique Record after
  *     merging.
  *
+ * - ProgresTrait:
```
Comment (smassung) [inline include/meta/util/multiway_merge.h]: typo

File: src/embeddings/tools/embedding_cooccur.cpp [hunk 4]
```diff
@@ -0,0 +1,81 @@
+/**
+ * @file embedding_coocur.cpp
```
Comment (smassung) [inline src/embeddings/tools/embedding_cooccur.cpp]: typo

File: src/embeddings/tools/glove.cpp [hunk 5]
```diff
@@ -176,13 +176,13 @@ class glove_trainer
             throw glove_exception{"no vocabulary file found in " + prefix};
         }
 
-        if (!filesystem::file_exists(prefix + "/coocur.bin"))
+        if (!filesystem::file_exists(prefix + "/cooccur.bin"))
         {
             LOG(fatal)
                 << "Coocurrence matrix has not yet been generated, please "
```
Comment (smassung) [inline src/embeddings/tools/glove.cpp]: typo

Comment (skystrife) [inline src/embeddings/tools/glove.cpp]: No?

Comment (smassung) [inline src/embeddings/tools/glove.cpp]: in the log text

File: include/meta/embeddings/cooccurrence_counter.h (hunk 2 above)
Comment (skystrife) [inline include/meta/embeddings/cooccurrence_counter.h]: We only want `it1 == it2` when `it1` and `it2` are both the "end" iterator. This is only true when `items_` is empty.

File: src/embeddings/tools/glove.cpp (hunk 5 above)
Comment (skystrife) [inline src/embeddings/tools/glove.cpp]: Ah, now I see it.

File: include/meta/embeddings/cooccurrence_counter.h (hunk 2 above)
Comment (smassung) [inline include/meta/embeddings/cooccurrence_counter.h]: Ah, I see.

Comment (smassung) [inline include/meta/embeddings/cooccurrence_counter.h]: Ah, I see.

File: src/embeddings/cooccurrence_counter.cpp [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+                {
+                    history.clear();
+                }
+                else if (tok == "</s>" && break_on_tags_)
+                {
+                    continue;
+                }
+                else
+                {
+                    // ignore out-of-vocabulary words
```
Comment (smassung) [inline src/embeddings/cooccurrence_counter.cpp]: (thought I left this comment already)

Add an option for dirty vs clean sampling? Not sure if it makes much difference.

Comment (skystrife) [inline src/embeddings/cooccurrence_counter.cpp]: Levy and Goldberg have this to say:

> **Deleting  Rare  Words  (`del`)** While  it  is  common to ignore words that are rare in the training corpus, word2vec removes  these  tokens  from the corpus before creating context windows.   As with subsampling, this variation narrows the distance between tokens, inserting new word-context pairs that did not exist in the original corpus with the same window size. Though this variation may also  have  an  effect  on  performance,  preliminary experiments  showed  that  it  was  small,  and  we therefore do not investigate its effect in this paper.

Not sure if I care about adding it. It adds a bunch of if/else into this loop to handle the case where `tid` and/or `*it` is not actually a value in the vocabulary.

**Comment by smassung** [comment]
Ready to merge? All looks good to me.

**Pull request merged**